### ISSUE TITLE
Twig_Template::getAttribute, array like objects and properties that equals null

### ISSUE THREAD

**Issue by Iwasawafag** [twigphp/Twig#1557]
Hello.

I'm using Twig library as view renderer with Yii 1.1.14 and whenever I'm trying to reference a property of an `ArrayAccess` object that equals `null`, uncaught exception is being thrown.

Here's what's going on: 

So, in template file I have this: `{{ model.foo | raw }}`, 
which is being compiled to something like this  `echo $this->getAttribute($_model_, "foo");`. 

Inside of `Twig_Template::getAttribute` method, `$model` is being checked for being an array, or an instance of `ArrayAccess` class and that requested key exists in object. But in case of ArrayAccess-like object it uses `isset()` function for this check, which is incorrect, I believe:

**/lib/Twig/Template.php:[358](https://github.com/twigphp/Twig/blob/v1.16.2/lib/Twig/Template.php#L358)**
```php
if ((is_array($object) && array_key_exists($arrayItem, $object))
    || ($object instanceof ArrayAccess && isset($object[$arrayItem]))
) {
    if ($isDefinedTest) {
        return true;
    }

    return $object[$arrayItem];
}
```

However, since `$model->foo` equals `null`, `isset()` returns `false`, like it supposed to, according to php documentation:

> **isset** — Determine if a variable is set and is not NULL

Magic method `__get` never gets called and so it falls through `getAttribute` function to the very end, where this block executes:

**/lib/Twig/Template.php:[462](https://github.com/twigphp/Twig/blob/v1.16.2/lib/Twig/Template.php#L462)**
```php
// Some objects throw exceptions when they have __call, and the method we try
// to call is not supported. If ignoreStrictCheck is true, we should return null.
try {
    $ret = call_user_func_array(array($object, $method), $arguments);
} catch (BadMethodCallException $e) {
    if ($call && ($ignoreStrictCheck || !$this->env->isStrictVariables())) {
        return;
    }
    throw $e;
}
```

and this is where I'm getting that exception from. Instead of `BadMethodCallException` Yii throws custom `CException` class (which is inherited from `Exception`), saying *{class} and its behaviors do not have a method or closure named "{name}"*, And so template rendering terminates because of this.

I'm not sure if it works as intended.

---------------------------

Because of `isset(...) === false` does not really means we can not access property, I think check for ArrayAccess objects should be changed to something like:

```php
if (is_array($object) && array_key_exists($arrayItem, $object)) {
    if ($isDefinedTest) {
        return true;
    }

    return $object[$arrayItem];
} elseif ($object instanceof ArrayAccess) {
    try {
        $value = $object[$arrayItem];

        if ($isDefinedTest) {
	        return true;
        }

        return $value;
    } catch (\Exception $e) {
    }
}
```

It looks more like a dirty hack, but I can't come up with more elegant way to check if property can be accessed.

**Comment by Tobion** [comment]
Should be closed as duplicate of #360

`isset` on an array access objects calls `offsetExists`. So it's up the the user to correctly return whatever he wants to be allowed to access. If `null` is allowed, he must change his implementation of offsetExists. Btw, if you `unset` this key and only set it to null, then it would be kinda bad as unsetting would not make the key disappear from `isset`.

**Comment by codedokode** [comment]
To me it looks like an error in the code. As `isset` cannot differentitate between null and missing property we should call `offsetExists()` directly and the code should look like 

```php

if ((is_array($object) && array_key_exists($arrayItem, $object))
    || ($object instanceof ArrayAccess && $object->offsetExists($arrayItem))
```

The property containing null and the missing property are not the same thing. Using `isset` is wrong here.
